metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -225,6 +272,32 @@ private ApiError updateFeature(
         if (featureName.equals(MetadataVersion.FEATURE_NAME)) {
             // Perform additional checks if we're updating metadata.version
             return updateMetadataVersion(newVersion, upgradeType.equals(FeatureUpdate.UpgradeType.UNSAFE_DOWNGRADE), records::add);
+        } else if (featureName.equals(KRaftVersion.FEATURE_NAME)) {
+            if (upgradeType.equals(FeatureUpdate.UpgradeType.UPGRADE)) {
+                try {
+                    kraftVersionAccessor.upgradeKRaftVersion(
+                        currentClaimedEpoch,
+                        KRaftVersion.fromFeatureLevel(newVersion),
+                        validateOnly
+                    );
+                    /* Add the noop record so that there is at least one offset to wait on to
+                     * complete the upgrade RPC
+                     */
+                    records.add(new ApiMessageAndVersion(new NoOpRecord(), (short) 0));
```

Comment:
we shouldn't add this if `validateOnly` is set, right?

Reply:
We could. I was trying to make it consistent with the rest of the code in this method. For other features, this code always adds an ApiMessageAndVersion to the records list. It is the caller of this method, `updateFeatures`, that returns an empty list if the RPC is meant to only validate.

Reply:
ok, in that case it looks good.